src/ray/common/asio/instrumented_io_context.cc

```diff
@@ -100,24 +72,63 @@ void instrumented_io_context::post(std::function<void()> handler,
   // the handlers to execute without segfaulting even if the thread that constructed the
   // io_context object exits without joining the handler threads (which can happen in a
   // hard exit).
-  boost::asio::io_context::post([handler = std::move(handler), stats = it->second,
-                                 global_stats = global_stats_, name, start]() {
-    int64_t start_execution = absl::GetCurrentTimeNanos();
-    // Execute actual handler.
-    handler();
-    int64_t end_execution = absl::GetCurrentTimeNanos();
-    // Update queue and execution time stats.
-    const auto queue_time_ns = start_execution - start;
-    const auto execution_time_ns = end_execution - start_execution;
-    // Update handler-specific stats.
-    {
-      absl::MutexLock lock(&(stats->mutex));
-      // Handler-specific execution stats.
-      stats->stats.cum_execution_time += execution_time_ns;
+  boost::asio::io_context::post(
+      [handler = std::move(handler), stats = start_and_stats.second,
+       global_stats = global_stats_, name, start = start_and_stats.first]() {
+        Instrument(std::move(handler), stats, global_stats, start);
+      });
+}
+
+std::pair<int64_t, std::shared_ptr<GuardedHandlerStats>> instrumented_io_context::Enqueue(
+    const std::string name) {
+  auto stats = GetOrCreate(name);
+  auto enqueue_time = Enqueue(stats);
+  return std::make_pair<int64_t, std::shared_ptr<GuardedHandlerStats>>(
+      std::move(enqueue_time), std::move(stats));
+}
+
+int64_t instrumented_io_context::Enqueue(std::shared_ptr<GuardedHandlerStats> stats) {
+  {
+    absl::MutexLock lock(&(stats->mutex));
+    stats->stats.cum_count++;
+    stats->stats.curr_count++;
+  }
+  return absl::GetCurrentTimeNanos();
+}
+
+std::shared_ptr<GuardedHandlerStats> instrumented_io_context::Instrument(
+    std::function<void()> fn, const std::string name) {
+  auto stats = GetOrCreate(name);
+  Instrument(std::move(fn), stats, global_stats_);
+  return stats;
+}
+
+void instrumented_io_context::Instrument(std::function<void()> fn,
```

Comment:
This is actually executing the fn right? So we should call it `RecordExecution` instead of instrument.